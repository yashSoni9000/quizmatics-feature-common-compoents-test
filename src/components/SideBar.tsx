import { useContext, useState } from "react";
import { NavLink } from "react-router-dom";

import {
    Box,
    List,
    ListItem,
    ListItemButton,
    ListItemText,
    Button,
} from "@mui/material";

import ArrowBackIosNewIcon from "@mui/icons-material/ArrowBackIosNew";

import { AppContext } from "@src/context/AppProvider";
import { Role } from "@src/ts/enums";
import {
    HOST_SIDEBAR_PATHS,
    PARTICIPANT_SIDEBAR_PATHS,
} from "@src/constants/routes";

interface Props {
    sideBarIsOpened: boolean;
    displaySideBar: (value: boolean) => void;
}

export default function SideBar({ sideBarIsOpened, displaySideBar }: Props) {
    const { role } = useContext(AppContext)?.userDetails || {};
    const [isSidebarCollapsed, setIsSidebarCollapsed] = useState(false);

    const routes =
        role === Role.Host ? HOST_SIDEBAR_PATHS : PARTICIPANT_SIDEBAR_PATHS;

    function handleOnClick() {
        displaySideBar(false);
    }

    function toggleSideBar() {
        setIsSidebarCollapsed((value) => !value);
    }

    return (
        <Box
            className={`h-full flex-col z-20 top-0 justify-between transition-all animate-slideIn lg:flex fixed lg:static w-60 ${
                isSidebarCollapsed && "lg:w-16"
            } ${sideBarIsOpened ? "flex" : "hidden"}`}
            sx={{
                backgroundColor: "background.default",
                "& .active > * > div": {
                    backgroundColor: "background.focusColor",
                    color: "primary.contrastText",
                },
            }}
            data-testid="sidebar"
        >
            <List>
                {routes.map(({ label, url, Icon }) => (
                    <NavLink key={url} end={url === ""} to={url}>
                        <ListItem disablePadding>
                            <ListItemButton
                                className="m-1 rounded-lg gap-4 h-12"
                                onClick={handleOnClick}
                            >
                                <Icon />
                                <ListItemText
                                    primary={label}
                                    className={`flex text-inherit pr-2 truncate ${
                                        isSidebarCollapsed && "lg:hidden"
                                    }`}
                                />
                            </ListItemButton>
                        </ListItem>
                    </NavLink>
                ))}
            </List>

            <Button
                className="ml-auto p-4 border-t border-gray-300 rounded-none border-solid w-full h-16 hidden lg:block"
                onClick={toggleSideBar}
                data-testid="collapse-button"
            >
                <ArrowBackIosNewIcon
                    className={`${
                        isSidebarCollapsed && "rotate-180"
                    } transition-all duration-300`}
                    sx={{ color: "background.contrastText" }}
                />
            </Button>
        </Box>
    );
}
