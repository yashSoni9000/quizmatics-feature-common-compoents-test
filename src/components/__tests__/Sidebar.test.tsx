import { fireEvent, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";

import SideBar from "../SideBar";
import { MemoryRouter } from "react-router-dom";

const mockedDisplaySidebar = jest.fn();
const mockedUsedLocation = jest.fn();

jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useLocation: mockedUsedLocation,
}));

describe("Sidebar Test", () => {
    it("snapshot test", () => {
        const { asFragment } = render(
            <MemoryRouter>
                <SideBar
                    sideBarIsOpened={true}
                    displaySideBar={mockedDisplaySidebar}
                />
            </MemoryRouter>,
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("Clicking on Arrow button collpase sidebar", () => {
        const { asFragment } = render(
            <MemoryRouter>
                <SideBar
                    sideBarIsOpened={true}
                    displaySideBar={mockedDisplaySidebar}
                />
            </MemoryRouter>,
        );

        const toggleButton = screen.getByTestId("collapse-button");
        const sideBar = screen.getByTestId("sidebar");

        expect(sideBar).toHaveClass("flex");
        fireEvent.click(toggleButton);
        expect(sideBar).toHaveClass("lg:w-16");

        fireEvent.click(toggleButton);
        expect(sideBar).toHaveClass("flex");

        expect(asFragment()).toMatchSnapshot();
    });

    it("Clicking on sidebar button set it to active button", () => {
        render(
            <MemoryRouter>
                <SideBar
                    sideBarIsOpened={true}
                    displaySideBar={mockedDisplaySidebar}
                />
            </MemoryRouter>,
        );

        const navLinkButtons = screen.getAllByRole("link");
        const navLinkButton = navLinkButtons.find(
            (button) => !button.classList.contains("active"),
        );

        const listItemButton = navLinkButton?.querySelector("[role=button]");

        if (!listItemButton) throw new Error();

        expect(navLinkButton).not.toHaveClass("active");
        fireEvent.click(listItemButton);

        expect(navLinkButton).toHaveClass("active");
        expect(mockedDisplaySidebar).toHaveBeenCalledWith(false);
    });
});
