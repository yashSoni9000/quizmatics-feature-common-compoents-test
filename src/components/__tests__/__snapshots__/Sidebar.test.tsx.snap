// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar Test Clicking on Arrow button collpase sidebar 1`] = `
<DocumentFragment>
  <div
    class="h-full flex-col z-20 top-0 justify-between transition-all animate-slideIn lg:flex fixed lg:static w-60 false flex MuiBox-root css-yodspd"
    data-testid="sidebar"
  >
    <ul
      class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
    >
      <a
        aria-current="page"
        class="active"
        href="/"
      >
        <li
          class="MuiListItem-root MuiListItem-gutters css-1nxmd3h-MuiListItem-root"
        >
          <div
            class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters m-1 rounded-lg gap-4 h-12 css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
            role="button"
            tabindex="0"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="SpaceDashboardIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M11 21H5c-1.1 0-2-.9-2-2V5c0-1.1.9-2 2-2h6v18zm2 0h6c1.1 0 2-.9 2-2v-7h-8v9zm8-11V5c0-1.1-.9-2-2-2h-6v7h8z"
              />
            </svg>
            <div
              class="MuiListItemText-root flex text-inherit pr-2 truncate false css-tlelie-MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
              >
                Dashboard
              </span>
            </div>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </div>
        </li>
      </a>
      <a
        class=""
        href="/groups"
      >
        <li
          class="MuiListItem-root MuiListItem-gutters css-1nxmd3h-MuiListItem-root"
        >
          <div
            class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters m-1 rounded-lg gap-4 h-12 css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
            role="button"
            tabindex="0"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="BookmarkIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M17 3H7c-1.1 0-1.99.9-1.99 2L5 21l7-3 7 3V5c0-1.1-.9-2-2-2z"
              />
            </svg>
            <div
              class="MuiListItemText-root flex text-inherit pr-2 truncate false css-tlelie-MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
              >
                My Groups
              </span>
            </div>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </div>
        </li>
      </a>
      <a
        class=""
        href="/quizzes"
      >
        <li
          class="MuiListItem-root MuiListItem-gutters css-1nxmd3h-MuiListItem-root"
        >
          <div
            class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters m-1 rounded-lg gap-4 h-12 css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
            role="button"
            tabindex="0"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="EventIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
              />
            </svg>
            <div
              class="MuiListItemText-root flex text-inherit pr-2 truncate false css-tlelie-MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
              >
                My Quizzes
              </span>
            </div>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </div>
        </li>
      </a>
      <a
        class=""
        href="/progress"
      >
        <li
          class="MuiListItem-root MuiListItem-gutters css-1nxmd3h-MuiListItem-root"
        >
          <div
            class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters m-1 rounded-lg gap-4 h-12 css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
            role="button"
            tabindex="0"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="AssessmentIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z"
              />
            </svg>
            <div
              class="MuiListItemText-root flex text-inherit pr-2 truncate false css-tlelie-MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
              >
                Progress Tracking
              </span>
            </div>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </div>
        </li>
      </a>
    </ul>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium ml-auto p-4 border-t border-gray-300 rounded-none border-solid w-full h-16 hidden lg:block css-1e6y48t-MuiButtonBase-root-MuiButton-root"
      data-testid="collapse-button"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium false transition-all duration-300 css-sfm0z2-MuiSvgIcon-root"
        data-testid="ArrowBackIosNewIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M17.77 3.77 16 2 6 12l10 10 1.77-1.77L9.54 12z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`Sidebar Test snapshot test 1`] = `
<DocumentFragment>
  <div
    class="h-full flex-col z-20 top-0 justify-between transition-all animate-slideIn lg:flex fixed lg:static w-60 false flex MuiBox-root css-yodspd"
    data-testid="sidebar"
  >
    <ul
      class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
    >
      <a
        aria-current="page"
        class="active"
        href="/"
      >
        <li
          class="MuiListItem-root MuiListItem-gutters css-1nxmd3h-MuiListItem-root"
        >
          <div
            class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters m-1 rounded-lg gap-4 h-12 css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
            role="button"
            tabindex="0"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="SpaceDashboardIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M11 21H5c-1.1 0-2-.9-2-2V5c0-1.1.9-2 2-2h6v18zm2 0h6c1.1 0 2-.9 2-2v-7h-8v9zm8-11V5c0-1.1-.9-2-2-2h-6v7h8z"
              />
            </svg>
            <div
              class="MuiListItemText-root flex text-inherit pr-2 truncate false css-tlelie-MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
              >
                Dashboard
              </span>
            </div>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </div>
        </li>
      </a>
      <a
        class=""
        href="/groups"
      >
        <li
          class="MuiListItem-root MuiListItem-gutters css-1nxmd3h-MuiListItem-root"
        >
          <div
            class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters m-1 rounded-lg gap-4 h-12 css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
            role="button"
            tabindex="0"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="BookmarkIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M17 3H7c-1.1 0-1.99.9-1.99 2L5 21l7-3 7 3V5c0-1.1-.9-2-2-2z"
              />
            </svg>
            <div
              class="MuiListItemText-root flex text-inherit pr-2 truncate false css-tlelie-MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
              >
                My Groups
              </span>
            </div>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </div>
        </li>
      </a>
      <a
        class=""
        href="/quizzes"
      >
        <li
          class="MuiListItem-root MuiListItem-gutters css-1nxmd3h-MuiListItem-root"
        >
          <div
            class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters m-1 rounded-lg gap-4 h-12 css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
            role="button"
            tabindex="0"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="EventIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
              />
            </svg>
            <div
              class="MuiListItemText-root flex text-inherit pr-2 truncate false css-tlelie-MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
              >
                My Quizzes
              </span>
            </div>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </div>
        </li>
      </a>
      <a
        class=""
        href="/progress"
      >
        <li
          class="MuiListItem-root MuiListItem-gutters css-1nxmd3h-MuiListItem-root"
        >
          <div
            class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters m-1 rounded-lg gap-4 h-12 css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
            role="button"
            tabindex="0"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="AssessmentIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z"
              />
            </svg>
            <div
              class="MuiListItemText-root flex text-inherit pr-2 truncate false css-tlelie-MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
              >
                Progress Tracking
              </span>
            </div>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </div>
        </li>
      </a>
    </ul>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium ml-auto p-4 border-t border-gray-300 rounded-none border-solid w-full h-16 hidden lg:block css-1e6y48t-MuiButtonBase-root-MuiButton-root"
      data-testid="collapse-button"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium false transition-all duration-300 css-sfm0z2-MuiSvgIcon-root"
        data-testid="ArrowBackIosNewIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M17.77 3.77 16 2 6 12l10 10 1.77-1.77L9.54 12z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;
